src/omv/ports/stm32/omv_gpu.c

```diff
@@ -0,0 +1,247 @@
+
+        if (!alpha_palette) {
+            if (!color_palette) {
+                for (int i = 0; i < 256; i++) {
+                    clut[i] = (0xff << 24) | COLOR_Y_TO_RGB888(i);
+                }
+            } else {
+                for (int i = 0; i < 256; i++) {
+                    int pixel = color_palette[i];
+                    clut[i] = (0xff << 24) |
+                              (COLOR_RGB565_TO_R8(pixel) << 16) |
+                              (COLOR_RGB565_TO_G8(pixel) << 8) |
+                              COLOR_RGB565_TO_B8(pixel);
+                }
+            }
+        } else {
+            if (!color_palette) {
+                for (int i = 0; i < 256; i++) {
+                    clut[i] = (alpha_palette[i] << 24) | COLOR_Y_TO_RGB888(i);
+                }
+            } else {
+                for (int i = 0; i < 256; i++) {
+                    int pixel = color_palette[i];
+                    clut[i] = (alpha_palette[i] << 24) |
+                              (COLOR_RGB565_TO_R8(pixel) << 16) |
+                              (COLOR_RGB565_TO_G8(pixel) << 8) |
+                              COLOR_RGB565_TO_B8(pixel);
+                }
+            }
+        }
+
+        DMA2D_CLUTCfgTypeDef cfg;
+        cfg.pCLUT = clut;
+        cfg.CLUTColorMode = DMA2D_CCM_ARGB8888;
+        cfg.Size = 255;
+        #if __DCACHE_PRESENT
+        SCB_CleanDCache_by_Addr(clut, 256 * sizeof(uint32_t));
+        #endif
+        HAL_DMA2D_CLUTLoad(&dma2d, cfg, 1);
+        HAL_DMA2D_PollForTransfer(&dma2d, 1000);
+    } else {
+        dma2d.LayerCfg[1].InputColorMode = DMA2D_INPUT_RGB565;
+        dma2d.LayerCfg[1].AlphaMode = DMA2D_REPLACE_ALPHA;
+    }
+
+    dma2d.LayerCfg[1].InputOffset = src_img->w - src_rect->w;
+    dma2d.LayerCfg[1].InputAlpha = (alpha * 255) / 256;
+    #if defined(MCU_SERIES_F7) || defined(MCU_SERIES_H7)
+    dma2d.LayerCfg[1].AlphaInverted = DMA2D_REGULAR_ALPHA;
+    dma2d.LayerCfg[1].RedBlueSwap = DMA2D_RB_REGULAR;
+    #endif
+    #if defined(MCU_SERIES_H7)
+    dma2d.LayerCfg[1].ChromaSubSampling = DMA2D_NO_CSS;
+    #endif
+
+    HAL_DMA2D_ConfigLayer(&dma2d, 1);
+
+    uint16_t *dst16 = IMAGE_COMPUTE_RGB565_PIXEL_ROW_PTR(dst_img, dst_rect->y) + dst_rect->x;
+
+    #if __DCACHE_PRESENT
+    // Ensures any cached writes to dst16 are flushed.
+    uint16_t *dst16_tmp = dst16;
+    for (int i = 0; i < dst_rect->h; i++) {
+        SCB_CleanInvalidateDCache_by_Addr(dst16_tmp, dst_rect->w * sizeof(uint16_t));
+        dst16_tmp += dst_img->w;
+    }
+    #endif
+
+    uint32_t src;
+
+    if (src_img->pixfmt == PIXFORMAT_GRAYSCALE) {
+        uint8_t *src8 = IMAGE_COMPUTE_GRAYSCALE_PIXEL_ROW_PTR(src_img, src_rect->y) + src_rect->x;
+        src = (uint32_t) src8;
+
+        #if __DCACHE_PRESENT
+        uint8_t *src8_tmp = src8;
+        for (int i = 0; i < src_rect->h; i++) {
+            SCB_CleanDCache_by_Addr(src8_tmp, src_rect->w);
+            src8_tmp += src_img->w;
+        }
```

Comment:
```suggestion
        for (int i = 0; i < src_rect->h; i++) {
            SCB_CleanDCache_by_Addr(src8_tmp, src_rect->w);
            src8_tmp += src_img->w;
        }
```
Why flush line by line ?

Reply:
You might only be drawing a sub-rectangle of the image. There's no need to clean everything.

Reply:
But it will be extremely slower in normal use cases. I don't care much about the DMA2D just pointing that out.